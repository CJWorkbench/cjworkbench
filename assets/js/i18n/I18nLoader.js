import React from 'react'
import { I18nProvider } from '@lingui/react'
import fetchCatalog from './catalogs'

export class I18nLoader extends React.Component {
  state = {
    catalogs: {},
  }
  
  loadCatalog = (language) => {
      this.setState(state => ({
          catalogs: {
            ...state.catalogs,
            [language]: fetchCatalog(language)
          }
      }))
  }
  
  componentDidMount() {
    this.loadCatalog(this.props.language)
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { language } = nextProps
    const { catalogs } = nextState

    if (language !== this.props.language && !catalogs[language]) {
      this.loadCatalog(language)
      return false
    }

    return true
  }

  render () {
    const { children, language } = this.props
    const { catalogs } = this.state

    // Skip rendering when catalog isn't loaded.
    if (!catalogs[language]) {
        return null
    }

    return (
      <I18nProvider language={language} catalogs={catalogs}>
        {children}
      </I18nProvider>
    )
  }
}
