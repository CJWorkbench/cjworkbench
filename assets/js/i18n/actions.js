const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
