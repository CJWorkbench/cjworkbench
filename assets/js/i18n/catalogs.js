import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
