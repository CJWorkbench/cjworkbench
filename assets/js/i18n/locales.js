import { t } from '@lingui/macro'

export const supportedLocales = {
  en: t('js.i18n.locales.en')`English`,
  el: t('js.i18n.locales.el')`Greek`
}

export const supportedLocaleIds = Object.keys(supportedLocales)

const defaultLocale = 'en'

export const currentLocale = window.i18nConfig && window.i18nConfig.locale && isSupported(window.i18nConfig.locale) ? window.i18nConfig.locale : defaultLocale

function isSupported (locale) {
  return Object.prototype.hasOwnProperty.call(supportedLocales, locale)
}
