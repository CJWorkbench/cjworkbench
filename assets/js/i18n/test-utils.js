import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
