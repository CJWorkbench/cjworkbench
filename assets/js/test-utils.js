/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
