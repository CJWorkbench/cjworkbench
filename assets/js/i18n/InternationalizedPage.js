import React from 'react'
import PropTypes from 'prop-types'
import { en, el } from 'make-plural/plurals'
import { i18n } from '@lingui/core'
import { I18nProvider } from '@lingui/react'
import { currentLocaleId } from './locales'
import Spinner from '../Spinner'

__webpack_public_path__ = window.STATIC_URL + 'bundles/' // eslint-disable-line

async function loadLocale (localeId) {
  const { messages } = await import(
    /* webpackChunkName: "[request]" */
    `@lingui/loader!../../locale/${localeId}/messages.po`
  )
  i18n.loadLocaleData({
    en: { plurals: en },
    el: { plurals: el }
  })
  i18n.load(localeId, messages)
  i18n.activate(localeId)
  return i18n
}

function useActivatedI18n (localeId) {
  const [i18n, setI18n] = React.useState(null)

  React.useEffect(() => {
    loadLocale(currentLocaleId).then(setI18n)
  }, [])

  return i18n
}

export default function InternationalizedPage (props) {
  const { children } = props
  const i18n = useActivatedI18n(currentLocaleId)

  if (i18n === null) {
    return <Spinner />
  }

  return (
    <I18nProvider i18n={i18n} forceRenderOnLocaleChange={false}>
      {children}
    </I18nProvider>
  )
}
InternationalizedPage.propTypes = {
  children: PropTypes.node.isRequired
}
