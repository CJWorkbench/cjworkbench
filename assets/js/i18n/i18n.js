import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
