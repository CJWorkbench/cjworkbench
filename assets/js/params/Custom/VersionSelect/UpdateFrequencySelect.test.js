/* globals afterEach, beforeEach, describe, expect, it, jest */
import { mockStore } from '../../../test-utils'
import React from 'react'
import ConnectedUpdateFrequencySelect, { UpdateFrequencySelect } from './UpdateFrequencySelect'
import { mountWithI18n } from '../../../i18n/test-utils'
import { Provider } from 'react-redux'

describe('UpdateFrequencySelect', () => {
  describe('shallow', () => {
    const defaultProps = {
      workflowId: 123,
      wfModuleId: 212,
      isReadOnly: false,
      isAnonymous: false,
      lastCheckDate: new Date(Date.parse('2018-05-28T19:00:54.154Z')),
      isAutofetch: false, // start in Manual mode
      isEmailUpdates: false,
      fetchInterval: 300,
      setEmailUpdates: jest.fn(),
      trySetAutofetch: jest.fn()
    }

    let dateSpy
    beforeEach(() => {
      dateSpy = jest.spyOn(Date, 'now').mockImplementation(() => 1527534812631)
    })
    afterEach(() => dateSpy.mockRestore())

    const wrapper = (extraProps) => {
      return mountWithI18n(
        <UpdateFrequencySelect
          {...defaultProps}
          {...extraProps}
        />
      )
    }

    it('matches snapshot', () => {
      expect(wrapper()).toMatchSnapshot()
    })

    it('does not render modal on first load', () => {
      expect(wrapper().find('UpdateFrequencySelectModal')).toHaveLength(0)
    })

    it('does not open modal when not read-only', () => {
      const w = wrapper({ isReadOnly: true })
      w.find('a[title="change auto-update settings"]').simulate('click')
      expect(w.find('UpdateFrequencySelectModal')).toHaveLength(0)
    })

    it('does not open modal when anonymous', () => {
      const w = wrapper({ isAnonymous: true })
      w.find('a[title="change auto-update settings"]').simulate('click')
      expect(w.find('UpdateFrequencySelectModal')).toHaveLength(0)
    })
  })

  describe('connected to state', () => {
    let wrapper = null
    afterEach(() => {
      if (wrapper) {
        wrapper.unmount()
        wrapper = null
      }
    })

    const sampleState = {
      workflow: {
        id: 123,
        read_only: false,
        is_anonymous: false,
        tab_slugs: ['tab-11', 'tab-12']
      },
      tabs: {
        'tab-11': { wf_modules: [1, 212] }
      },
      wfModules: {
        1: { id: 1, tab_slug: 'tab-11', name: 'Ignore this one' },
        212: { id: 212, tab_slug: 'tab-11', auto_update_data: true, update_interval: 3600, update_units: 'days', notifications: false, last_update_check: '2018-05-28T19:00:54.154141Z' }
      }
    }

    it('should get settings from state', () => {
      const store = {
        getState: () => sampleState,
        dispatch: jest.fn(),
        subscribe: jest.fn()
      }
      wrapper = mountWithI18n(
        <Provider store={store}>
          <ConnectedUpdateFrequencySelect
            wfModuleId={212}
            lastCheckDate={null}
          />
        </Provider>
      )
      expect(wrapper.find('time').prop('dateTime')).toEqual('2018-05-28T19:00:54.154Z')
      wrapper.find('a[title="change auto-update settings"]').simulate('click')
      const modal = wrapper.find('UpdateFrequencySelectModal')
      expect(modal.prop('fetchInterval')).toBe(3600)
    })

    it('should not crash on a placeholder', () => {
      // can this even happen?
      const store = {
        getState: () => ({
          workflow: { id: 123, read_only: false, is_anonymous: false, wf_modules: ['nonce_212'] },
          wfModules: {}
        }),
        dispatch: jest.fn(),
        subscribe: jest.fn()
      }
      wrapper = mountWithI18n(
        <Provider store={store}>
          <ConnectedUpdateFrequencySelect
            wfModuleId={212}
          />
        </Provider>
      )
      wrapper.find('a[title="change auto-update settings"]').simulate('click')
      expect(true).toBe(true)
    })

    it('should set autofetch (calling API method)', () => {
      const api = {
        trySetWfModuleAutofetch: jest.fn(() => Promise.resolve({ isAutofetch: true, fetchInterval: 7200 }))
      }
      const store = mockStore(sampleState, api)
      wrapper = mountWithI18n(
        <Provider store={store}>
          <ConnectedUpdateFrequencySelect
            wfModuleId={212}
            lastCheckDate={null}
          />
        </Provider>
      )
      wrapper.find('a[title="change auto-update settings"]').simulate('click')
      const modal = wrapper.find('UpdateFrequencySelectModal')
      modal.prop('trySetAutofetch')(true, 7600)
      expect(api.trySetWfModuleAutofetch).toHaveBeenCalledWith(212, true, 7600)
    })

    it('should dispatch setNotifications (and call the API method)', () => {
      const api = {
        setWfModuleNotifications: jest.fn(() => Promise.resolve(null))
      }
      const store = mockStore(sampleState, api)
      wrapper = mountWithI18n(
        <Provider store={store}>
          <ConnectedUpdateFrequencySelect
            wfModuleId={212}
            lastCheckDate={null}
          />
        </Provider>
      )
      wrapper.find('a[title="change auto-update settings"]').simulate('click')
      const modal = wrapper.find('UpdateFrequencySelectModal')
      modal.prop('setEmailUpdates')(false)
      expect(api.setWfModuleNotifications).toHaveBeenCalledWith(212, false)
    })
  })
})
