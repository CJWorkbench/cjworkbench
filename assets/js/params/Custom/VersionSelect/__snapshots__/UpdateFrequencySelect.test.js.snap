// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateFrequencySelect shallow matches snapshot 1`] = `
<WithI18n
  fetchInterval={300}
  i18n={
    Object {
      "_": [Function],
      "_activeLanguageData": Object {
        "plurals": [Function],
      },
      "_activeMessages": Object {},
      "_catalogs": Object {
        "en": Object {
          "languageData": Object {},
          "messages": Object {},
        },
      },
      "_dev": Object {
        "compile": [Function],
        "loadLanguageData": [Function],
      },
      "_language": "en",
      "_locales": undefined,
      "_missing": undefined,
      "plural": [Function],
      "select": [Function],
      "selectOrdinal": [Function],
      "t": [Function],
    }
  }
  isAnonymous={false}
  isAutofetch={false}
  isEmailUpdates={false}
  isReadOnly={false}
  lastCheckDate={2018-05-28T19:00:54.154Z}
  setEmailUpdates={[MockFunction]}
  trySetAutofetch={[MockFunction]}
  wfModuleId={212}
  workflowId={123}
>
  <I18n
    update={true}
    withHash={true}
  >
    <UpdateFrequencySelect
      fetchInterval={300}
      i18n={
        Object {
          "_": [Function],
          "_activeLanguageData": Object {
            "plurals": [Function],
          },
          "_activeMessages": Object {},
          "_catalogs": Object {
            "en": Object {
              "languageData": Object {},
              "messages": Object {},
            },
          },
          "_dev": Object {
            "compile": [Function],
            "loadLanguageData": [Function],
          },
          "_language": "en",
          "_locales": undefined,
          "_missing": undefined,
          "plural": [Function],
          "select": [Function],
          "selectOrdinal": [Function],
          "t": [Function],
        }
      }
      isAnonymous={false}
      isAutofetch={false}
      isEmailUpdates={false}
      isReadOnly={false}
      lastCheckDate={2018-05-28T19:00:54.154Z}
      setEmailUpdates={[MockFunction]}
      trySetAutofetch={[MockFunction]}
      wfModuleId={212}
      workflowId={123}
    >
      <div
        className="update-frequency-select"
      >
        <div
          className="update-option"
        >
          <span
            className="version-box-option"
          >
            <WithI18n
              defaults="Update"
              id="js.params.Custom.VersionSelect.UpdateFrequencySelect.update"
            >
              <I18n
                update={true}
                withHash={true}
              >
                <Trans
                  defaults="Update"
                  i18n={
                    Object {
                      "_": [Function],
                      "_activeLanguageData": Object {
                        "plurals": [Function],
                      },
                      "_activeMessages": Object {},
                      "_catalogs": Object {
                        "en": Object {
                          "languageData": Object {},
                          "messages": Object {},
                        },
                      },
                      "_dev": Object {
                        "compile": [Function],
                        "loadLanguageData": [Function],
                      },
                      "_language": "en",
                      "_locales": undefined,
                      "_missing": undefined,
                      "plural": [Function],
                      "select": [Function],
                      "selectOrdinal": [Function],
                      "t": [Function],
                    }
                  }
                  id="js.params.Custom.VersionSelect.UpdateFrequencySelect.update"
                >
                  <Render
                    value="js.params.Custom.VersionSelect.UpdateFrequencySelect.update"
                  >
                    js.params.Custom.VersionSelect.UpdateFrequencySelect.update
                  </Render>
                </Trans>
              </I18n>
            </WithI18n>
             
          </span>
          <a
            className="content-1 ml-1 action-link"
            href="#"
            onClick={[Function]}
            title="change auto-update settings"
          >
            <WithI18n
              defaults="Manual"
              id="js.params.Custom.VersionSelect.UpdateFrequencySelect.manual"
            >
              <I18n
                update={true}
                withHash={true}
              >
                <Trans
                  defaults="Manual"
                  i18n={
                    Object {
                      "_": [Function],
                      "_activeLanguageData": Object {
                        "plurals": [Function],
                      },
                      "_activeMessages": Object {},
                      "_catalogs": Object {
                        "en": Object {
                          "languageData": Object {},
                          "messages": Object {},
                        },
                      },
                      "_dev": Object {
                        "compile": [Function],
                        "loadLanguageData": [Function],
                      },
                      "_language": "en",
                      "_locales": undefined,
                      "_missing": undefined,
                      "plural": [Function],
                      "select": [Function],
                      "selectOrdinal": [Function],
                      "t": [Function],
                    }
                  }
                  id="js.params.Custom.VersionSelect.UpdateFrequencySelect.manual"
                >
                  <Render
                    value="js.params.Custom.VersionSelect.UpdateFrequencySelect.manual"
                  >
                    js.params.Custom.VersionSelect.UpdateFrequencySelect.manual
                  </Render>
                </Trans>
              </I18n>
            </WithI18n>
          </a>
        </div>
        <div
          className="last-checked"
        >
          <WithI18n
            components={
              Array [
                <time
                  dateTime="2018-05-28T19:00:54.154Z"
                />,
              ]
            }
            defaults="Checked <0>{0}</0>"
            description="The parameter is a time difference (i.e. something like '4h ago'. The tag is a <time> tag."
            id="js.params.Custom.VersionSelect.UpdateFrequencySelect.lastChecked"
            values={
              Object {
                "0": "{minutes}m ago",
              }
            }
          >
            <I18n
              update={true}
              withHash={true}
            >
              <Trans
                components={
                  Array [
                    <time
                      dateTime="2018-05-28T19:00:54.154Z"
                    />,
                  ]
                }
                defaults="Checked <0>{0}</0>"
                description="The parameter is a time difference (i.e. something like '4h ago'. The tag is a <time> tag."
                i18n={
                  Object {
                    "_": [Function],
                    "_activeLanguageData": Object {
                      "plurals": [Function],
                    },
                    "_activeMessages": Object {},
                    "_catalogs": Object {
                      "en": Object {
                        "languageData": Object {},
                        "messages": Object {},
                      },
                    },
                    "_dev": Object {
                      "compile": [Function],
                      "loadLanguageData": [Function],
                    },
                    "_language": "en",
                    "_locales": undefined,
                    "_missing": undefined,
                    "plural": [Function],
                    "select": [Function],
                    "selectOrdinal": [Function],
                    "t": [Function],
                  }
                }
                id="js.params.Custom.VersionSelect.UpdateFrequencySelect.lastChecked"
                values={
                  Object {
                    "0": "{minutes}m ago",
                  }
                }
              >
                <Render
                  value="js.params.Custom.VersionSelect.UpdateFrequencySelect.lastChecked"
                >
                  js.params.Custom.VersionSelect.UpdateFrequencySelect.lastChecked
                </Render>
              </Trans>
            </I18n>
          </WithI18n>
        </div>
      </div>
    </UpdateFrequencySelect>
  </I18n>
</WithI18n>
`;
